In [3]:
from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


In [5]:
import numpy as np
import tensorflow as tf
from tensorflow.keras import layers, models
import json
from sklearn.cluster import KMeans
from itertools import product
import os

# --- Neural Feature Extractor ---
def create_feature_extractor(input_shape=(30, 30, 1)):
    """Creates a CNN model for feature extraction from grids"""
    model = models.Sequential([
        layers.Conv2D(32, (3, 3), activation='relu', input_shape=input_shape),
        layers.MaxPooling2D((2, 2)),
        layers.Conv2D(64, (3, 3), activation='relu'),
        layers.MaxPooling2D((2, 2)),
        layers.Conv2D(128, (3, 3), activation='relu'),
        layers.Flatten(),
        layers.Dense(256, activation='relu'),
        layers.Dropout(0.3),
        layers.Dense(128, name='feature_vector')
    ])
    return model

# --- Symbolic Program Generator ---
PRIMITIVES = [
    'rotate', 'mirrorlr', 'mirrorud', 'lcrop', 'rcrop', 'ucrop', 'dcrop',
    'recolor', 'select', 'fill', 'overlay', 'resize'
]

def generate_programs(max_length=3):
    """Generates symbolic programs of primitive operations"""
    programs = []
    for length in range(1, max_length + 1):
        for combo in product(PRIMITIVES, repeat=length):
            programs.append(list(combo))
    return programs

# --- Differentiable Program Executor ---
class ProgramExecutor(tf.keras.Model):
    """Differentiable executor for symbolic programs"""
    def __init__(self, feature_extractor, num_primitives):
        super().__init__()
        self.feature_extractor = feature_extractor
        self.primitive_weights = layers.Dense(num_primitives, activation='softmax')

    def call(self, inputs):
        # Extract features from input/output pairs
        input_grid, output_grid = inputs
        input_feat = self.feature_extractor(input_grid)
        output_feat = self.feature_extractor(output_grid)

        # Compute primitive probabilities
        combined = tf.concat([input_feat, output_feat], axis=-1)
        return self.primitive_weights(combined)

# --- Neural-Symbolic Solver ---
class NeuralSymbolicSolver:
    def __init__(self):
        self.feature_extractor = create_feature_extractor()
        self.executor = ProgramExecutor(self.feature_extractor, len(PRIMITIVES))
        self.executor.compile(
            optimizer='adam',
            loss='categorical_crossentropy'
        )
        self.programs = generate_programs(max_length=3)

    def train(self, train_data, epochs=10, batch_size=32):
        """Train on demonstration pairs"""
        # Prepare training data
        X_in, X_out, y_primitive = [], [], []

        for example in train_data:
            input_grid = self.preprocess(example['input'])
            output_grid = self.preprocess(example['output'])

            # Find best primitive (simplified for example)
            best_primitive = self.find_best_primitive(input_grid, output_grid)

            X_in.append(input_grid)
            X_out.append(output_grid)
            y_primitive.append(PRIMITIVES.index(best_primitive))

        # Train the model
        X_in = np.array(X_in)
        X_out = np.array(X_out)
        y_primitive = tf.keras.utils.to_categorical(y_primitive, len(PRIMITIVES))

        self.executor.fit(
            [X_in, X_out], y_primitive,
            epochs=epochs,
            batch_size=batch_size
        )

    def solve(self, input_grid, output_grid):
        """Solve a new problem using neural-guided program synthesis"""
        # Predict primitive probabilities
        input_pp = self.preprocess(input_grid)
        output_pp = self.preprocess(output_grid)
        primitive_probs = self.executor.predict(
            [np.array([input_pp]), np.array([output_pp])]
        )[0]

        # Rank programs by primitive probabilities
        program_scores = []
        for program in self.programs:
            score = np.prod([primitive_probs[PRIMITIVES.index(p)] for p in program])
            program_scores.append((program, score))

        # Try top programs
        program_scores.sort(key=lambda x: x[1], reverse=True)

        for program, _ in program_scores[:10]:  # Try top 10
            result = self.execute_program(input_grid, program)
            if np.array_equal(result, output_grid):
                return program

        return None  # No solution found

    def execute_program(self, grid, program):
        """Execute a symbolic program on a grid"""
        current = np.array(grid)
        for op in program:
            if op == 'rotate':
                current = np.rot90(current, k=-1)
            elif op == 'mirrorlr':
                current = np.fliplr(current)
            elif op == 'mirrorud':
                current = np.flipud(current)
            elif op == 'lcrop':
                current = current[:, 1:] if current.shape[1] > 1 else current
            elif op == 'rcrop':
                current = current[:, :-1] if current.shape[1] > 1 else current
            elif op == 'ucrop':
                current = current[1:, :] if current.shape[0] > 1 else current
            elif op == 'dcrop':
                current = current[:-1, :] if current.shape[0] > 1 else current
            elif op == 'recolor':
                current = self.learn_recoloring(current)
            # Additional operations would be implemented here
        return current

    def learn_recoloring(self, grid):
        """Learn color mapping using clustering (simplified)"""
        # In practice, this would compare input/output colors
        return grid  # Placeholder

    def find_best_primitive(self, input_grid, output_grid):
        """Find best matching primitive (simplified heuristic)"""
        # In practice, use neural features for this
        for primitive in ['rotate', 'mirrorlr', 'mirrorud']:
            transformed = self.execute_program(input_grid, [primitive])
            if np.array_equal(transformed, output_grid):
                return primitive
        return 'recolor'  # Default

    def preprocess(self, grid, size=30):
        """Preprocess grid to fixed size with padding"""
        h, w = len(grid), len(grid[0])
        padded = np.zeros((size, size), dtype=int)
        padded[:h, :w] = grid
        return np.expand_dims(padded, axis=-1)

# --- Main Execution ---
if __name__ == "__main__":
    # Load training data
    file_path = 'drive/MyDrive/Colab Notebooks/arc-prize-2025/arc-prize-2025/arc-agi_training_challenges.json'
    if not os.path.exists(file_path):
        print(f"Error: File not found at {file_path}")
        print("Please ensure your Google Drive is mounted correctly and the file path is correct.")
        print("To mount your Drive, you can use the following code in a new cell:")
        print("from google.colab import drive\ndrive.mount('/content/drive')")
    else:
        with open(file_path, 'r') as f:
            train_data = json.load(f)

        # Prepare training examples
        training_examples = []
        for case_id, case_data in train_data.items():
            for example in case_data['train']:
                training_examples.append({
                    'input': example['input'],
                    'output': example['output']
                })

        # Initialize and train solver
        solver = NeuralSymbolicSolver()
        solver.train(training_examples[:100], epochs=5, batch_size=16)

        # Test on a sample case
        sample_case = list(train_data.values())[0]
        input_grid = sample_case['train'][0]['input']
        output_grid = sample_case['train'][0]['output']

        solution = solver.solve(input_grid, output_grid)
        print(f"Solution: {solution}")

        # Visualize results
        if solution:
            result = solver.execute_program(input_grid, solution)
            print("Original Input:")
            print(np.array(input_grid))
            print("\nSolved Output:")
            print(result)
            print("\nTarget Output:")
            print(np.array(output_grid))

Epoch 1/5
7/7 ━━━━━━━━━━━━━━━━━━━━ 3s 64ms/step - loss: 1.3795
Epoch 2/5
7/7 ━━━━━━━━━━━━━━━━━━━━ 1s 61ms/step - loss: 1.4348e-04
Epoch 3/5
7/7 ━━━━━━━━━━━━━━━━━━━━ 1s 65ms/step - loss: 0.0000e+00
Epoch 4/5
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 35ms/step - loss: 0.0000e+00
Epoch 5/5
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 38ms/step - loss: 0.0000e+00
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 157ms/step
Solution: None


# Task
Fix the provided code so that it can find a solution to the ARC task.

## Enhance `find best primitive`

### Subtask:
Enhance the `find_best_primitive` function to be more robust by considering all primitives and using a similarity metric to find the best one.


**Reasoning**:
I will now modify the `find_best_primitive` function to iterate through all primitives, apply them, and calculate the negative mean squared error to find the best-matching primitive, as outlined in the instructions. This will make the primitive selection process more robust and less reliant on a fixed, hardcoded order. I will also need to modify the `execute_program` function to handle single primitive operations correctly.



In [6]:
def find_best_primitive(self, input_grid, output_grid):
    """Find best matching primitive using a similarity metric."""
    best_primitive = None
    max_similarity = -np.inf

    for primitive in PRIMITIVES:
        transformed_grid = self.execute_program(np.array(input_grid), [primitive])

        # Ensure transformed_grid and output_grid have the same shape for comparison
        h_out, w_out = output_grid.shape
        h_trans, w_trans = transformed_grid.shape

        padded_transformed = np.zeros_like(output_grid)

        min_h = min(h_out, h_trans)
        min_w = min(w_out, w_trans)

        padded_transformed[:min_h, :min_w] = transformed_grid[:min_h, :min_w]

        # Using negative mean squared error as similarity
        similarity = -np.mean((padded_transformed - output_grid) ** 2)

        if similarity > max_similarity:
            max_similarity = similarity
            best_primitive = primitive

    return best_primitive

NeuralSymbolicSolver.find_best_primitive = find_best_primitive

# We need to ensure the `execute_program` function can handle the numpy array format
def execute_program(self, grid, program):
    """Execute a symbolic program on a grid"""
    current = np.array(grid)
    for op in program:
        if op == 'rotate':
            current = np.rot90(current, k=-1)
        elif op == 'mirrorlr':
            current = np.fliplr(current)
        elif op == 'mirrorud':
            current = np.flipud(current)
        elif op == 'lcrop':
            current = current[:, 1:] if current.shape[1] > 1 else current
        elif op == 'rcrop':
            current = current[:, :-1] if current.shape[1] > 1 else current
        elif op == 'ucrop':
            current = current[1:, :] if current.shape[0] > 1 else current
        elif op == 'dcrop':
            current = current[:-1, :] if current.shape[0] > 1 else current
        elif op == 'recolor':
            current = self.learn_recoloring(current)
        # Additional operations would be implemented here
    return current

NeuralSymbolicSolver.execute_program = execute_program

## Expand program search

### Subtask:
Increase the number of programs the solver attempts from the top 10 to a larger number (e.g., top 50 or 100).


**Reasoning**:
I need to modify the `solve` method in the `NeuralSymbolicSolver` class to change the number of programs attempted from the top 10 to the top 50.



In [7]:
class NeuralSymbolicSolver:
    def __init__(self):
        self.feature_extractor = create_feature_extractor()
        self.executor = ProgramExecutor(self.feature_extractor, len(PRIMITIVES))
        self.executor.compile(
            optimizer='adam',
            loss='categorical_crossentropy'
        )
        self.programs = generate_programs(max_length=3)

    def train(self, train_data, epochs=10, batch_size=32):
        """Train on demonstration pairs"""
        # Prepare training data
        X_in, X_out, y_primitive = [], [], []

        for example in train_data:
            input_grid = self.preprocess(example['input'])
            output_grid = self.preprocess(example['output'])

            # Find best primitive (simplified for example)
            best_primitive = self.find_best_primitive(np.array(example['input']), np.array(example['output'])) # Ensure numpy arrays are passed

            X_in.append(input_grid)
            X_out.append(output_grid)
            y_primitive.append(PRIMITIVES.index(best_primitive))

        # Train the model
        X_in = np.array(X_in)
        X_out = np.array(X_out)
        y_primitive = tf.keras.utils.to_categorical(y_primitive, len(PRIMITIVES))

        self.executor.fit(
            [X_in, X_out], y_primitive,
            epochs=epochs,
            batch_size=batch_size
        )

    def solve(self, input_grid, output_grid):
        """Solve a new problem using neural-guided program synthesis"""
        # Predict primitive probabilities
        input_pp = self.preprocess(input_grid)
        output_pp = self.preprocess(output_grid)
        primitive_probs = self.executor.predict(
            [np.array([input_pp]), np.array([output_pp])]
        )[0]

        # Rank programs by primitive probabilities
        program_scores = []
        for program in self.programs:
            score = np.prod([primitive_probs[PRIMITIVES.index(p)] for p in program])
            program_scores.append((program, score))

        # Try top programs
        program_scores.sort(key=lambda x: x[1], reverse=True)

        for program, _ in program_scores[:50]:  # Try top 50 instead of 10
            result = self.execute_program(input_grid, program)
            if np.array_equal(result, output_grid):
                return program

        return None  # No solution found

    def execute_program(self, grid, program):
        """Execute a symbolic program on a grid"""
        current = np.array(grid)
        for op in program:
            if op == 'rotate':
                current = np.rot90(current, k=-1)
            elif op == 'mirrorlr':
                current = np.fliplr(current)
            elif op == 'mirrorud':
                current = np.flipud(current)
            elif op == 'lcrop':
                current = current[:, 1:] if current.shape[1] > 1 else current
            elif op == 'rcrop':
                current = current[:, :-1] if current.shape[1] > 1 else current
            elif op == 'ucrop':
                current = current[1:, :] if current.shape[0] > 1 else current
            elif op == 'dcrop':
                current = current[:-1:, :] if current.shape[0] > 1 else current # Fixed typo here
            elif op == 'recolor':
                current = self.learn_recoloring(current)
            # Additional operations would be implemented here
        return current

    def learn_recoloring(self, grid):
        """Learn color mapping using clustering (simplified)"""
        # In practice, this would compare input/output colors
        return grid  # Placeholder

    def find_best_primitive(self, input_grid, output_grid):
        """Find best matching primitive using a similarity metric."""
        best_primitive = None
        max_similarity = -np.inf

        for primitive in PRIMITIVES:
            transformed_grid = self.execute_program(np.array(input_grid), [primitive])

            # Ensure transformed_grid and output_grid have the same shape for comparison
            h_out, w_out = output_grid.shape
            h_trans, w_trans = transformed_grid.shape

            # Calculate padding needed
            pad_h = max(0, h_out - h_trans)
            pad_w = max(0, w_out - w_trans)

            # Pad the transformed grid
            padded_transformed = np.pad(transformed_grid, ((0, pad_h), (0, pad_w)), mode='constant', constant_values=0)

            # Trim padded_transformed if it's larger than output_grid (shouldn't happen with padding logic, but as a safeguard)
            padded_transformed = padded_transformed[:h_out, :w_out]


            # Using negative mean squared error as similarity
            similarity = -np.mean((padded_transformed - output_grid) ** 2)

            if similarity > max_similarity:
                max_similarity = similarity
                best_primitive = primitive

        return best_primitive


    def preprocess(self, grid, size=30):
        """Preprocess grid to fixed size with padding"""
        h, w = len(grid), len(grid[0])
        padded = np.zeros((size, size), dtype=int)
        padded[:h, :w] = grid
        return np.expand_dims(padded, axis=-1)

# Re-instantiate the solver with the modified class definition
solver = NeuralSymbolicSolver()

# Train the solver again with the updated class
# Prepare training examples
training_examples = []
for case_id, case_data in train_data.items():
    for example in case_data['train']:
        training_examples.append({
            'input': example['input'],
            'output': example['output']
        })

solver.train(training_examples[:100], epochs=5, batch_size=16)

# Test on a sample case
sample_case = list(train_data.values())[0]
input_grid = sample_case['train'][0]['input']
output_grid = sample_case['train'][0]['output']

solution = solver.solve(input_grid, output_grid)
print(f"Solution: {solution}")

# Visualize results
if solution:
    result = solver.execute_program(input_grid, solution)
    print("Original Input:")
    print(np.array(input_grid))
    print("\nSolved Output:")
    print(result)
    print("\nTarget Output:")
    print(np.array(output_grid))


/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/5
7/7 ━━━━━━━━━━━━━━━━━━━━ 4s 39ms/step - loss: 2.3115
Epoch 2/5
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 36ms/step - loss: 1.7963
Epoch 3/5
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 37ms/step - loss: 1.6187
Epoch 4/5
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 35ms/step - loss: 1.5310
Epoch 5/5
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 39ms/step - loss: 1.1784
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 154ms/step
Solution: None


## Improve training

### Subtask:
Modify the training process to use a larger portion of the training data and train for more epochs to allow the model to learn more effectively.


**Reasoning**:
Modify the main execution block to use a larger portion of the training data and increase the number of epochs for training.



In [8]:
# Re-instantiate the solver with the current class definition
solver = NeuralSymbolicSolver()

# Prepare training examples
training_examples = []
for case_id, case_data in train_data.items():
    for example in case_data['train']:
        training_examples.append({
            'input': example['input'],
            'output': example['output']
        })

# Train the solver again with increased data and epochs
solver.train(training_examples[:500], epochs=15, batch_size=16) # Increased data to 500 and epochs to 15

# Test on a sample case
sample_case = list(train_data.values())[0]
input_grid = sample_case['train'][0]['input']
output_grid = sample_case['train'][0]['output']

solution = solver.solve(input_grid, output_grid)
print(f"Solution: {solution}")

# Visualize results
if solution:
    result = solver.execute_program(input_grid, solution)
    print("Original Input:")
    print(np.array(input_grid))
    print("\nSolved Output:")
    print(result)
    print("\nTarget Output:")
    print(np.array(output_grid))

/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 5s 66ms/step - loss: 2.2227
Epoch 2/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 2s 42ms/step - loss: 1.8473
Epoch 3/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 1s 35ms/step - loss: 1.5790
Epoch 4/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 1s 35ms/step - loss: 1.4356
Epoch 5/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 1s 35ms/step - loss: 1.3597
Epoch 6/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 1s 35ms/step - loss: 1.1557
Epoch 7/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 1s 35ms/step - loss: 0.9853
Epoch 8/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 1s 35ms/step - loss: 0.9432
Epoch 9/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 1s 36ms/step - loss: 0.7012
Epoch 10/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 1s 44ms/step - loss: 0.6926
Epoch 11/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 3s 56ms/step - loss: 0.7189
Epoch 12/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 1s 35ms/step - loss: 0.5866
Epoch 13/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 1s 37ms/step - loss: 0.3938
Epoch 14/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 1s 35ms/step - loss: 0.4311
Epoch 15/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 1s 35ms/step - loss: 0.2720
1/1 

## Refine feature extractor

### Subtask:
Refine the feature extractor architecture to better capture the salient features of the input grids.


**Reasoning**:
I will redefine the `create_feature_extractor` function to include `BatchNormalization` layers after each `Conv2D` and `Dense` layer, increase the number of filters in the `Conv2D` layers to 64, 128, and 256, and increase the number of units in the final `Dense` layer to 256, as per the instructions. Then, I will re-run the training and solving pipeline to evaluate the impact of these changes.



In [9]:
def create_feature_extractor(input_shape=(30, 30, 1)):
    """Creates a CNN model for feature extraction from grids"""
    model = models.Sequential([
        layers.Conv2D(64, (3, 3), activation='relu', input_shape=input_shape),
        layers.BatchNormalization(),
        layers.MaxPooling2D((2, 2)),
        layers.Conv2D(128, (3, 3), activation='relu'),
        layers.BatchNormalization(),
        layers.MaxPooling2D((2, 2)),
        layers.Conv2D(256, (3, 3), activation='relu'),
        layers.BatchNormalization(),
        layers.Flatten(),
        layers.Dense(256, activation='relu'),
        layers.BatchNormalization(),
        layers.Dropout(0.3),
        layers.Dense(256, name='feature_vector')
    ])
    return model

# Re-instantiate the solver with the new feature extractor
solver = NeuralSymbolicSolver()

# Prepare training examples
training_examples = []
for case_id, case_data in train_data.items():
    for example in case_data['train']:
        training_examples.append({
            'input': example['input'],
            'output': example['output']
        })

# Train the solver again with the new feature extractor
solver.train(training_examples[:500], epochs=15, batch_size=16)

# Test on a sample case
sample_case = list(train_data.values())[0]
input_grid = sample_case['train'][0]['input']
output_grid = sample_case['train'][0]['output']

solution = solver.solve(input_grid, output_grid)
print(f"Solution: {solution}")

# Visualize results
if solution:
    result = solver.execute_program(input_grid, solution)
    print("Original Input:")
    print(np.array(input_grid))
    print("\nSolved Output:")
    print(result)
    print("\nTarget Output:")
    print(np.array(output_grid))

/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 9s 150ms/step - loss: 2.2246
Epoch 2/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 4s 105ms/step - loss: 1.8758
Epoch 3/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 3s 105ms/step - loss: 1.6460
Epoch 4/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 3s 104ms/step - loss: 1.4437
Epoch 5/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 5s 157ms/step - loss: 1.3109
Epoch 6/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 3s 104ms/step - loss: 1.1844
Epoch 7/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 3s 104ms/step - loss: 0.9679
Epoch 8/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 7s 155ms/step - loss: 0.7718
Epoch 9/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 4s 105ms/step - loss: 0.6470
Epoch 10/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 3s 104ms/step - loss: 0.5944
Epoch 11/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 3s 106ms/step - loss: 0.4237
Epoch 12/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 5s 116ms/step - loss: 0.5849
Epoch 13/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 3s 105ms/step - loss: 0.3003
Epoch 14/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 3s 105ms/step - loss: 0.2336
Epoch 15/15
32/32 ━━━━━━━━━━━━━━━━━━━━ 6s 131ms/step - lo

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 288ms/step
Solution: None


## Summary:

### Data Analysis Key Findings
- The `find_best_primitive` function was enhanced to iterate through all available primitives and use negative mean squared error as a similarity metric to find the best one.
- The number of top programs to try was increased from 10 to 50 to broaden the search for a solution.
- The model was trained with more data (500 examples instead of 100) and for more epochs (15 instead of 5) to improve its learning. The training loss decreased from around 2.2 to approximately 0.27, indicating that the model was learning from the additional data.
- The feature extractor was refined by adding `BatchNormalization` layers, increasing the number of filters in the convolutional layers, and increasing the number of units in the final dense layer to capture more complex features.
- Despite all the enhancements made to the solver, including a more robust primitive selection, an expanded program search, improved training, and a refined feature extractor, the solver consistently failed to find a solution to the sample ARC task.

### Insights or Next Steps
- The failure to find a solution suggests that the current set of primitives in the Domain-Specific Language (DSL) may be insufficient to solve the task. It's recommended to explore adding more complex or domain-specific primitives.
- The current program synthesis approach relies on a simple search over ranked programs. A more sophisticated search algorithm, such as a beam search or a genetic algorithm, could be more effective at exploring the program space and finding a solution.
